sphinxcontrib/datatemplates/directive.py

```diff
@@ -114,3 +61,63 @@ def run(self):
         node.document = self.state.document
         nested_parse_with_titles(self.state, result, node)
         return node.children
+
+
+class DataTemplateJSON(DataTemplateBase):
+    def _load_data(self, env, data_source):
+        rel_filename, filename = env.relfn2path(data_source)
+        with open(filename, 'r') as f:
+            return json.load(f)
+
+
+def _handle_dialect_option(argument):
+    return rst.directives.choice(argument, ["auto"] + csv.list_dialects())
+
+
+class DataTemplateCSV(DataTemplateBase):
+    option_spec = dict(
+        DataTemplateBase.option_spec, **{
+            'headers': rst.directives.flag,
+            'dialect': _handle_dialect_option,
+        })
+
+    def _load_data(self, env, data_source):
+        rel_filename, filename = env.relfn2path(data_source)
+        with open(filename, 'r', newline='') as f:
+            dialect = self.options.get('dialect')
+            if dialect == "auto":
+                sample = f.read(8192)
+                f.seek(0)
+                sniffer = csv.Sniffer()
+                dialect = sniffer.sniff(sample)
+            if 'headers' in self.options:
+                if dialect is None:
+                    r = csv.DictReader(f)
+                else:
+                    r = csv.DictReader(f, dialect=dialect)
+            else:
+                if dialect is None:
+                    r = csv.reader(f)
+                else:
+                    r = csv.reader(f, dialect=dialect)
+            return list(r)
+
+
+class DataTemplateYAML(DataTemplateBase):
+    def _load_data(self, env, data_source):
+        rel_filename, filename = env.relfn2path(data_source)
+        with open(filename, 'r') as f:
+            return yaml.safe_load(f)
+
+
+class DataTemplateXML(DataTemplateBase):
+    def _load_data(self, env, data_source):
+        rel_filename, filename = env.relfn2path(data_source)
+        return ET.parse(filename).getroot()
+
+
+class DataTemplateLegacy(DataTemplateYAML):
+    def run(self):
+        LOG.warning('Using the datatemplate directive is deprecated. '
```

Comment:
This is good. It will appear in the output and break builds where the flag to treat warnings as errors is enabled.